src/data/indicator_function.lean

```diff
@@ -82,19 +82,53 @@ lemma indicator_add (s : set α) (f g : α → β) :
   indicator s (λa, f a + g a) = λa, indicator s f a + indicator s g a :=
 by { funext, simp only [indicator], split_ifs, { refl }, rw add_zero }
 
-lemma indicator_smul {𝕜 : Type*} [monoid 𝕜] [distrib_mul_action 𝕜 β] (s : set α) (r : 𝕜) (f : α → β) :
+variables (β)
+instance is_add_monoid_hom.indicator (s : set α) : is_add_monoid_hom (λf:α → β, indicator s f) :=
+{ map_add := λ _ _, indicator_add _ _ _,
+  map_zero := indicator_zero _ _ }
+
+variables {β} {𝕜 : Type*} [monoid 𝕜] [distrib_mul_action 𝕜 β]
+
+lemma indicator_smul (s : set α) (r : 𝕜) (f : α → β) :
   indicator s (λ (x : α), r • f x) = λ (x : α), r • indicator s f x :=
 by { simp only [indicator], funext, split_ifs, refl, exact (smul_zero r).symm }
 
-lemma indicator_neg {β : Type*} [add_group β] (s : set α) (f : α → β) :
-  indicator s (λa, - f a) = λa, - indicator s f a :=
-by { funext, simp only [indicator], split_ifs, { refl }, rw neg_zero }
+end add_monoid
+
+section add_group
+variables [add_group β] {s t : set α} {f g : α → β} {a : α}
 
-lemma indicator_sub {β : Type*} [add_group β] (s : set α) (f g : α → β) :
+variables (β)
+instance is_add_group_hom.indicator (s : set α) : is_add_group_hom (λf:α → β, indicator s f) :=
+{ .. is_add_monoid_hom.indicator β s }
+variables {β}
+
+lemma indicator_neg (s : set α) (f : α → β) : indicator s (λa, - f a) = λa, - indicator s f a :=
+show indicator s (- f) = - indicator s f, from is_add_group_hom.map_neg _ _
+
+lemma indicator_sub (s : set α) (f g : α → β) :
   indicator s (λa, f a - g a) = λa, indicator s f a - indicator s g a :=
-by { funext, simp only [indicator], split_ifs, { refl }, rw sub_zero }
+show indicator s (f - g) = indicator s f - indicator s g, from is_add_group_hom.map_sub _ _ _
+
+lemma indicator_compl (s : set α) (f : α → β) : indicator (-s) f = λ a, f a - indicator s f a :=
+begin
+  funext,
+  simp only [indicator],
+  split_ifs with h₁ h₂,
+  { rw sub_zero },
+  { rw sub_self },
+  { rw ← mem_compl_iff at h₂, contradiction }
+end
+
+lemma indicator_sum {β} [add_comm_monoid β] {ι : Type*} (I : finset ι) (s : set α) (f : ι → α → β) :
+  indicator s (I.sum f) = I.sum (λ i, indicator s (f i)) :=
+begin
+  convert (finset.sum_hom _ _).symm,
+  split,
```

Comment:
Some typeclass problem makes `is_add_monoid_hom.indicator` not usable here. So I have to prove that `is_add_monoid_hom (indicator s)` again. 

Also, I have no idea how `map_add` is proved automatically. It is like magic.

Reply:
The following works:
```lean
  -- Lean fails to unify `add_comm_monoid.to_add_monoid (pi.add_comm_monoid ...)`
  -- with `pi.add_monoid ...`, so we use `convert`
  convert is_add_monoid_hom.indicator β s
```

Reply:
Asked Daniel [here](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/typeclass.20problems).